client_plugin/drivers/shared/shared_driver.go

```diff
@@ -402,8 +403,175 @@ func (d *VolumeDriver) Path(r volume.Request) volume.Response {
+	return d.processMount(r)
+}
+
+// processMount -  process a mount request
+func (d *VolumeDriver) processMount(r volume.MountRequest) volume.Response {
+	d.MountIDtoName[r.ID] = r.Name
+
+	// If the volume is already mounted , just increase the refcount.
+	// Note: for new keys, GO maps return zero value, so no need for if_exists.
+	refcnt := d.IncrRefCount(r.Name) // save map traversal
+	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
+	if refcnt > 1 {
+		log.WithFields(
+			log.Fields{"name": r.Name, "refcount": refcnt},
+		).Info("Already mounted, skipping mount. ")
+		return volume.Response{Mountpoint: d.GetMountPoint(r.Name)}
+	}
+
+	if plugin_utils.AlreadyMounted(r.Name, d.MountRoot) {
+		log.WithFields(log.Fields{"name": r.Name}).Info("Already mounted, skipping mount. ")
+		return volume.Response{Mountpoint: d.GetMountPoint(r.Name)}
+	}
+
+	mountpoint, err := d.MountVolume(r.Name, "", "", false, true)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": r.Name,
+				"error": err},
+		).Error("Failed to mount ")
+
+		refcnt, _ := d.DecrRefCount(r.Name)
+		if refcnt == 0 {
+			log.Infof("Detaching %s - it is not used anymore", r.Name)
+			// TODO: umount here
+		}
+		return volume.Response{Err: err.Error()}
+	}
+
+	return volume.Response{Mountpoint: mountpoint}
+}
+
+// MountVolume - Request attach and then mounts the volume.
+func (d *VolumeDriver) MountVolume(name string, fstype string, id string, isReadOnly bool, skipAttach bool) (string, error) {
+	mountpoint := d.GetMountPoint(name)
+	// First, make sure  that mountpoint exists.
+	err := fs.Mkdir(mountpoint)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": name,
+				"dir": mountpoint},
+		).Error("Failed to make directory for volume mount ")
+		return mountpoint, err
+	}
+
+	// Increase GRef
+	log.Infof("Before AtomicIncr")
+	err = d.kvStore.AtomicIncr(kvstore.VolPrefixGRef + name)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": name,
+				"error": err},
+		).Error("Failed to increase global refcount when processMount ")
+		return "", err
+	}
+
+	// Blocking wait until the state of volume becomes Mounted
+	// the change of global refcount will trigger one watcher on manager nodes
+	// watchers should start event handler to transit the state of volumes
+	info, err := d.kvStore.BlockingWaitAndGet(kvstore.VolPrefixState+name,
+		string(kvstore.VolStateMounted), kvstore.VolPrefixInfo+name)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to blocking wait for Mounted state. Error: %v.", err)
+		err = d.kvStore.AtomicDecr(kvstore.VolPrefixGRef + name)
+		if err != nil {
+			msg += fmt.Sprintf(" Also failed to decrease global refcount. Error: %v.", err)
+		}
+		log.WithFields(
+			log.Fields{"name": name,
+				"error": msg}).Error("")
+		return "", errors.New(msg)
+	}
+
+	// Start mounting
+	log.Infof("Volume state mounted, prepare to mounting locally")
+	var volRecord VolumeMetadata
+	// Unmarshal Info key
+	err = json.Unmarshal([]byte(info), &volRecord)
```

Comment:
With only this PR, there is no problem to achieve username and password.
I think what you observed here is because you re-write the info value with the marshal string from sharedVolConnectivityData, which only has port and service name. As a result username and password is overwritten. I added a comment in your PR regarding to this problem: https://github.com/vmware/docker-volume-vsphere/pull/1742/files#diff-d20300e752878bf03d64ac84f8f56dc5R368